### ISSUE TITLE
le9 and 5.15

### ISSUE THREAD

**Issue by hakavlad** [zen-kernel/zen-kernel#256]
Seems like le9 patch is not in 5.15-zen. Why?

**Comment by heftig** [comment]
Waiting for a port of MG-LRU.

**Comment by hakavlad** [comment]
But le9 can be applied without MG-LRU. You're not canceling any other patches because MG-LRU hasn't been released yet, are you?

**Comment by hakavlad** [comment]
@yuzhaogoogle Do you plan to release mg-lru for 5.15?

**Comment by damentz** [comment]
@hakavlad I rebased 5.14/lru-v5 to v5.15 on branch 5.15/lru, it's in 5.15/master as of now: https://github.com/zen-kernel/zen-kernel/commit/ffa18c91b1f1b2b084c8bd90d29df475d0e26552

Can you test and make sure it works?  So far I've only build tested to make sure there's no errors at build-time.

**Comment by sirlucjan** [comment]
Could you add [1] to the 5.15/prjc branch?

[1] https://github.com/zen-kernel/zen-kernel/commit/c2912e232e6373baff6804959b6288e2f3e5b75d

**Comment by damentz** [comment]
@sirlucjan done, also prefixed the description with **ZEN:** so I don't miss it next time: https://github.com/zen-kernel/zen-kernel/commit/09b9774798f91c4747ee3266e910eb8aebf2118c

**Comment by heftig** [comment]
@yuzhaogoogle Is there are newer patch series for multi-generational lru available? Apparently we have "v5", while the last submission to linux-mm was "v4"?

**Comment by damentz** [comment]
@heftig v5 was my internal naming, since I went through a few combination of patches.  The latest lru-v5 branch in the 5.14 branches is based on the v4 patch + le9,  That's also the same branch that I used for 5.15/lru.

**Comment by heftig** [comment]
@damentz Applying the v4 series + le9 + our tweaks results in different code, though. So I was wondering.

**Comment by yuzhaogoogle** [comment]
@heftig sorry for falling behind the schedule -- I meant to post v5 a couple weeks ago when 5.15 was just released. Will be posting a new version in the next few hours.

@damentz thanks for rebasing -- just took a look and Zen 5.15 has the best tested version:

- [MariaDB 20%+ faster](https://patchwork.kernel.org/project/linux-mm/cover/20210818063107.2696454-1-yuzhao@google.com/#24506153)
- [Memcached 40%+ faster](https://patchwork.kernel.org/project/linux-mm/cover/20210818063107.2696454-1-yuzhao@google.com/#24541807)
- [Apache Spark 10%+ faster](https://patchwork.kernel.org/project/linux-mm/cover/20210818063107.2696454-1-yuzhao@google.com/#24566713)
- [MongoDB 20%+ faster](https://patchwork.kernel.org/project/linux-mm/cover/20210818063107.2696454-1-yuzhao@google.com/#24581971)

I'll let you know once the new version I'm posting has become as well tested. (It would be for 5.16 by then.)

**Comment by yuzhaogoogle** [comment]
FYI: v5 posted [here](https://patchwork.kernel.org/project/linux-mm/cover/20211111041510.402534-1-yuzhao@google.com/)

No action needed for Zen 5.15.

**Comment by hakavlad** [comment]
>Can you test and make sure it works?

@damentz I tested zen 5.15/master. Build warnings: https://pastebin.com/Mc11fdbq

MG-LRU: OK. min_ttl_ms by default: 4000.

le9: OK. vm.cleam_low_kbytes=0 by default.

There are no more remarks. No problems found.

**Comment by heftig** [comment]
> FYI: v5 posted [here](https://patchwork.kernel.org/project/linux-mm/cover/20211111041510.402534-1-yuzhao@google.com/)
> 
> No action needed for Zen 5.15.

Are you sure? That series is quite different from what's in ZEN. It wouldn't be difficult to replace.

**Comment by hakavlad** [comment]
I set
```
/* Protect the working set accessed within the last N milliseconds. */
#ifdef CONFIG_ZEN_INTERACTIVE
static unsigned long lru_gen_min_ttl = 3000;
#else
```
Result:
```
$ cat /sys/kernel/mm/lru_gen/min_ttl_ms
12000
```
¯\_(ツ)_/¯

**Comment by damentz** [comment]
Ok, considering 5.14 isn't EOL yet and we're only on 5.15.1, I replaced the 5.15/lru branch with @yuzhaogoogle's v5 patch series and le9 on top.

In master, reverted original 5.15/lru merge (https://github.com/zen-kernel/zen-kernel/commit/d272c12260ecbd8971e12c28f999fc9612140eec), and merged new version (https://github.com/zen-kernel/zen-kernel/commit/d4c4863a013fc3c265c90e3fceaf2408078c30d7)

With that, a successful build should mean this is resolved.  I'll leave the issue open for another day or two in-case we made a mistake, but otherwise, I think we're done here.  Thanks everyone!

**Comment by yuzhaogoogle** [comment]
> Are you sure? That series is quite different from what's in ZEN. It wouldn't be difficult to replace.

The version in Zen 5.14 was well tested and benchmarked by our partners. The new series I posted for 5.15 hasn't gone through the same process yet. But if you are interested in having the new series in Zen 5.15, I'd be glad to provide any assistance if Zen users report any problems. Thanks!

**Comment by yuzhaogoogle** [comment]
> I set
> 
> ```
> /* Protect the working set accessed within the last N milliseconds. */
> #ifdef CONFIG_ZEN_INTERACTIVE
> static unsigned long lru_gen_min_ttl = 3000;
> #else
> ```
> 
> Result:
> 
> ```
> $ cat /sys/kernel/mm/lru_gen/min_ttl_ms
> 12000
> ```
> 
> ¯_(ツ)_/¯

Hmm, this is strange. What's your CONFIG_HZ?

**Comment by heftig** [comment]
I think it means 8d4b19a722ecf532ffe459e4f2556c4801135357 is wrong because the variable is in jiffies, not ms, so it's only correct with `HZ=1000`. With `HZ=250`, 3000 jiffies are 12 seconds. 

I think we should be using `msecs_to_jiffies(1000)` instead of a bare `1000`.

**Comment by hakavlad** [comment]
@yuzhaogoogle `CONFIG_HZ=250`, full: https://pastebin.com/kR841gYb

**Comment by heftig** [comment]
Whoops, `msecs_to_jiffies` is a function. `HZ` should do, then.

**Comment by heftig** [comment]
@yuzhaogoogle Also means the comment (`/* Protect the working set accessed within the last N milliseconds. */`) for the variable is wrong and should mention `jiffies` instead of `milliseconds`.

**Comment by damentz** [comment]
> > Are you sure? That series is quite different from what's in ZEN. It wouldn't be difficult to replace.
> 
> The version in Zen 5.14 was well tested and benchmarked by our partners. The new series I posted for 5.15 hasn't gone through the same process yet. But if you are interested in having the new series in Zen 5.15, I'd be glad to provide any assistance if Zen users report any problems. Thanks!

Yep, for the most part we want to run the latest version unless there's a known issue.  Sometimes this does introduce regressions (early adoption of futex2 comes to mind), but the payoff is usually worth the short term hassle.

> @yuzhaogoogle Also means the comment (/* Protect the working set accessed within the last N milliseconds. */) for the variable is wrong and should mention jiffies instead of milliseconds.

Considering `linux-zen` and `linux-lqx` both use 1000hz, this issue is technically benign.  Even if the actual value used is different than what was specified at boot, the behavior of MG-LRU is still directionally correct, protecting recent file cache.

Ok, other than that, I I think we're good here!  Marking as closed.

**Comment by yuzhaogoogle** [comment]
Yes, I'm on it. Will have it ready soon.

**Comment by yuzhaogoogle** [comment]
> @yuzhaogoogle have you got any plans on rebasing mglru on top of v5.16-rc1 please?

For 5.16: https://lore.kernel.org/linux-mm/20220104202227.2903605-1-yuzhao@google.com/

**Comment by pfactum** [comment]
> For 5.16: https://lore.kernel.org/linux-mm/20220104202227.2903605-1-yuzhao@google.com/

Thank you. I'm already using it for a couple of days, and things look good.

**Comment by yuzhaogoogle** [comment]
> > For 5.16: https://lore.kernel.org/linux-mm/20220104202227.2903605-1-yuzhao@google.com/
> 
> Thank you. I'm already using it for a couple of days, and things look good.

Glad to hear. I'm trying to get this patchset merged in 5.17. Feel free to report your user experience -- IMO, desktop users' voice have been falling on deaf ears for too long. We are large in numbers and we need to turn the table.

There is an instruction on how to do this at the end of [this link](https://lore.kernel.org/linux-mm/YdiKVJlClB3h1Kmg@google.com/).
